java/google/registry/rde/XjcToDomainResourceConverter.java

```diff
@@ -0,0 +1,329 @@
+    private final Registry tld;
+
+    GracePeriodConverter(XjcRdeDomain domain, Key<BillingEvent.Recurring> autoRenewBillingEvent) {
+      this.domain = domain;
+      this.autoRenewBillingEvent = autoRenewBillingEvent;
+      this.tld = Registry.get(
+          Registries.findTldForNameOrThrow(InternetDomainName.from(domain.getName())).toString());
+    }
+
+    @Override
+    public GracePeriod apply(XjcRgpStatusType gracePeriodStatus) {
+      // TODO: (wolfgang) address these items in code review:
+      // verify that this logic is correct
+      switch(gracePeriodStatus.getS()) {
+        case ADD_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.ADD,
+              domain.getCrDate().plus(this.tld.getAddGracePeriodLength()),
+              domain.getCrRr().getClient());
+        case AUTO_RENEW_PERIOD:
+          return GracePeriod.createForRecurring(
+              GracePeriodStatus.AUTO_RENEW,
+              domain.getUpDate().plus(this.tld.getAutoRenewGracePeriodLength()),
+              domain.getClID(),
+              autoRenewBillingEvent);
+        case PENDING_DELETE:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.PENDING_DELETE,
+              domain.getUpDate().plus(this.tld.getPendingDeleteLength()),
+              domain.getClID());
+        case REDEMPTION_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.REDEMPTION,
+              domain.getUpDate().plus(this.tld.getRedemptionGracePeriodLength()),
+              domain.getClID());
+        case RENEW_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.RENEW,
+              domain.getUpDate().plus(this.tld.getRenewGracePeriodLength()),
+              domain.getClID());
+        case TRANSFER_PERIOD:
+          return GracePeriod.createWithoutBillingEvent(
+              GracePeriodStatus.TRANSFER,
+              domain.getUpDate().plus(this.tld.getTransferGracePeriodLength()),
+              domain.getTrnData().getReRr().getValue());
+        default:
+          throw new IllegalArgumentException(
+              "Unsupported grace period status: " + gracePeriodStatus.getS());
+      }
+    }
+  }
+
+  /** Converts {@link XjcRdeDomain} to {@link DomainResource}. */
+  static DomainResource convertDomain(XjcRdeDomain domain) {
+    // First create history entry and autorenew billing event/poll message
+    // Autorenew billing event is required for creating AUTO_RENEW grace period
+    HistoryEntry historyEntry = createHistoryEntry(domain);
+    BillingEvent.Recurring autoRenewBillingEvent = createAutoRenewBillingEvent(domain, historyEntry);
+    createAutoRenewPollMessage(domain, historyEntry);
+    GracePeriodConverter gracePeriodConverter = new GracePeriodConverter(
+        domain, Key.create(autoRenewBillingEvent));
+    DomainResource.Builder builder = new DomainResource.Builder()
```

Comment:
You may want a check before this line that a domain with the given ROID doesn't already exist.  Without that, you could end up accidentally overwriting an existing domain that already has that ROID.